### PR TITLE
 #57: Configurable Documenten API Upload fields

### PR DESCRIPTION
**Pull request by Klaas-Ritense**

### CODE REVIEW

File: projects/valtimo/components/src/lib/services/valtimo-modal.service.ts [hunk 1]
```diff
@@ -29,7 +30,9 @@ export class ValtimoModalService {
   }
 
   get documentDefinitionName$(): Observable<string> {
-    return this._documentDefinitionName$.asObservable();
+    return this._documentDefinitionName$
+      .asObservable()
```
Comment (teodor-ritense) [inline projects/valtimo/components/src/lib/services/valtimo-modal.service.ts]: Remove asObservable, pipe automatically converts to observable

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.html [hunk 2]
```diff
@@ -80,6 +82,35 @@
...
+    [hideDescription]="!obs.formFields?.beschrijving?.visible"
+    [hideDocumentTitle]="!obs.formFields?.titel?.visible"
+    [hideDocumentType]="!obs.formFields?.informatieobjecttype?.visible"
+    [hideFilename]="!obs.formFields?.bestandsnaam?.visible"
+    [hideLanguage]="!obs.formFields?.taal?.visible"
+    [hideStatus]="!obs.formFields?.status?.visible"
+    [hideTags]="!obs.formFields?.trefwoorden?.visible"
+    [language]="obs.formFields?.taal?.defaultValue"
+    [status]="obs.formFields?.status?.defaultValue"
+    [tags]="obs.metadata?.trefwoorden"
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.html]: This is way too many Inputs, convert them into two different objects: config and prefillData. Config has all the 'hide' inputs and prefillData has the prefilled Data

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts [hunk 3]
```diff
@@ -198,6 +198,31 @@ export class DossierDetailTabDocumentenApiDocumentsComponent implements OnInit,
...
+      formFields.forEach(
+        formField =>
+          (obj[formField.key] = {
+            defaultValue: formField.defaultValue,
+            visible: formField.visible,
+            readonly: formField.readonly,
+          })
+      );
+      return obj;
+    })
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts]:
```suggestion
    map(formFields => formFields.reduce(
      (acc, curr) => ({
        ...acc,
        [curr.key]: {
          defaultValue: curr.defaultValue,
          visible: curr.visible,
          readonly: curr.readonly,
        },
      }),
      {}
    )
      );
      return obj;
    })
```

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-metadata-modal/documenten-api-metadata-modal.component.ts
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-metadata-modal/documenten-api-metadata-modal.component.ts:1]: Already added a comment on the inputs in the template

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.html [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        key === 'bestandsnaam' ||
+        key === 'titel' ||
+        key === 'beschrijving' ||
+        key === 'trefwoorden'
+      ) {
+        <cds-label>
+          {{ 'zgw.uploadFields.fields.defaultValue' | translate }}
+          <input cdsText formControlName="defaultValue" />
+        </cds-label>
+      }
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.html:88]: Use @switch

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.scss [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+.toggle-group {
+  display: flex;
+  flex-direction: column;
+  margin-top: 1rem;
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.scss]:
```suggestion
  margin-top: 16px;
```

File: projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.html [hunk 6]
```diff
@@ -81,8 +82,20 @@
...
+  [hideDocumentTitle]="hideDocumentTitle"
+  [hideDocumentType]="hideDocumentType"
+  [hideFilename]="hideFilename"
+  [hideLanguage]="hideLanguage"
+  [hideStatus]="hideStatus"
+  [hideTags]="hideTags"
   [language]="language"
   [status]="status"
   [supportsTrefwoorden]="(supportedDocumentenApiFeatures$ | async)?.supportsTrefwoorden"
+  [tags]="tags"
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.html]: Same thing as before

File: projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.ts [hunk 7]
```diff
@@ -169,4 +187,8 @@ export class DocumentenApiUploaderComponent
       )
       .subscribe();
   }
+
+  get tags(): string[] {
```
Comment (teodor-ritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.ts]:
```suggestion
  public get tags(): string[] {
```

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.html (hunk 4 above)
Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-upload-field-model/documenten-api-upload-field-modal.component.html:88]: Unfortunately, the `@Switch` cant do this nicely:
```
if (
        key === 'filename' ||
        key === 'bestandsnaam' ||
        key === 'titel' ||
        key === 'beschrijving' ||
        key === 'trefwoorden'
      )
```

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.html (hunk 2 above)
Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.html]: I tried to reduce it a bit

File: projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.html (hunk 6 above)
Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/formio/documenten-api-uploader/documenten-api-uploader.component.html]: changed

File: projects/valtimo/config/assets/core/nl.json [hunk 8]
```diff
@@ -934,9 +934,51 @@
         "valueMaxLength": "Trefwoord is te lang. Maximum lengte is {{ maxlength }} tekens."
       }
     },
+    "uploadFields": {
+      "uploadField": "Upload veld",
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
      "uploadField": "Uploadveld",
```

File: projects/valtimo/config/assets/core/nl.json [hunk 9]
```diff
@@ -934,9 +934,51 @@
...
+      "add": "Upload veld toevoegen",
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
      "add": "Uploadveld toevoegen",
```

File: projects/valtimo/config/assets/core/nl.json [hunk 10]
```diff
@@ -934,9 +934,51 @@
...
+      "deleteText": "Weet je zeker dat je dit upload veld wilt verwijderen?",
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
      "deleteText": "Weet je zeker dat je dit uploadveld wilt verwijderen?",
```

File: projects/valtimo/config/assets/core/nl.json [hunk 11]
```diff
@@ -934,9 +934,51 @@
...
+      "deleteMultipleText": "Weet je zeker dat je deze upload velden wilt verwijderen?",
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
      "deleteMultipleText": "Weet je zeker dat je deze uploadvelden wilt verwijderen?",
```

File: projects/valtimo/config/assets/core/nl.json [hunk 12]
```diff
@@ -934,9 +934,51 @@
...
+      "addModal": {
+        "header": "Upload veld toevoegen"
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
        "header": "Uploadveld toevoegen"
```

File: projects/valtimo/config/assets/core/nl.json [hunk 13]
```diff
@@ -934,9 +934,51 @@
...
+      },
+      "editModal": {
+        "header": "Upload veld bewerken"
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
        "header": "Uploadveld bewerken"
```

File: projects/valtimo/config/assets/core/nl.json [hunk 14]
```diff
@@ -934,9 +934,51 @@
...
+      },
+      "list": {
+        "noResultsTitle": "Geen upload velden",
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
        "noResultsTitle": "Geen uploadvelden",
```

File: projects/valtimo/config/assets/core/nl.json [hunk 15]
```diff
@@ -934,9 +934,51 @@
...
+        "noResultsDescription": "Voeg een upload veld toe om te beginnen"
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
        "noResultsDescription": "Voeg een uploadveld toe om te beginnen"
```

File: projects/valtimo/config/assets/core/nl.json [hunk 16]
```diff
@@ -934,9 +934,51 @@
...
+        "ontvangstdatum": "Ontvangstdatum",
+        "aanvullendeDatum": "AanvullendeDatum",
+        "trefwoorden": "Trefwoorden"
+      }
+    },
     "tabs": {
       "documentColumns": "Document kolommen",
-      "documentTags": "Document trefwoorden"
+      "documentTags": "Document trefwoorden",
+      "documentUploadFields": "Document upload velden"
```
Comment (mbritense) [inline projects/valtimo/config/assets/core/nl.json]:
```suggestion
      "documentUploadFields": "Document uploadvelden"
```

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts [hunk 17]
```diff
@@ -198,6 +198,49 @@ export class DossierDetailTabDocumentenApiDocumentsComponent implements OnInit,
     return this._sort$.pipe(map(sortValue => this.getSortStateFromSortString(sortValue?.sort)));
   }
 
+  public uploadFields$: Observable<any> = this.documentId$.pipe(
```
Comment (mbritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts]: Don't use Observable<any>, use strong typing

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts [hunk 18]
```diff
@@ -198,6 +198,49 @@ export class DossierDetailTabDocumentenApiDocumentsComponent implements OnInit,
...
+            visible: curr.visible,
+            readonly: curr.readonly,
+          },
+        }),
+        {}
+      )
+    )
+  );
+
+  public defaultValues$: Observable<{}> = this.uploadFields$.pipe(
```
Comment (mbritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts]: Use strong typing on observable

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-metadata-modal/documenten-api-metadata-modal.component.ts [hunk 19]
```diff
@@ -579,27 +557,53 @@ export class DocumentenApiMetadataModalComponent implements OnInit, OnChanges, O
     this._subscriptions.add(
       combineLatest([this.file$, this.userEmail$])
         .pipe(
+          take(1),
```
Comment (mbritense) [inline projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-metadata-modal/documenten-api-metadata-modal.component.ts]: What's the point of adding a take(1) of observable to subscriptions? Are you sure adding this take(1) does not break previous functionality?

Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-metadata-modal/documenten-api-metadata-modal.component.ts]: I've removed it

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts (hunk 17 above)
Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts]: change

File: projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts (hunk 18 above)
Comment (Klaas-Ritense) [reply projects/valtimo/zgw/src/lib/modules/documenten-api/components/documenten-api-documents/documenten-api-documents.component.ts]: changed

**Review by mbritense — APPROVED** [review]

**Pull request merged**